src/keydra/providers/gitlab.py

```diff
@@ -0,0 +1,175 @@
+            )
+
+        except Exception as e:  # pragma: no cover
+            LOGGER.warn('Failed to distribute secret to Gitlab repo {}: {}'.format(
+                config['repository'], e))
+            raise DistributionException(e)
+
+        LOGGER.info(
+            'Successfully distributed {}.{}.key to '
+            'gitlab.{}.{} '.format(
+                secret['provider'],
+                secret['key'],
+                config['repository'],
+                dest['key']
+            )
+        )
+
+        return dest
+
+    @classmethod
+    def _validate_repository_spec(cls, spec):
+        if 'repository' not in spec['config']:
+            return False, 'Attribute "config.repository" not present'
+
+        return True, ''
+
+    def _distribute(self, secret, destination):
+        try:
+            # We only support repo level distribution right now
+            return self._distribute_repository_secret(secret, destination)
+
+        except Exception as e:
+            raise DistributionException(e)
+
+    def _load_as_dict_from_repo(self, file_path, repo_name, repo_branch, filetype):
+        LOGGER.info(
+            'Loading remote file from GitLab repo {} at path {}'.format(repo_name, file_path)
+        )
+
+        file_content = self.repo_client.fetch_file_from_repository(
+            file_path=file_path,
+            repo_name=repo_name,
+            repo_branch=repo_branch
+        )
+
+        LOGGER.debug('Remote file content: \n{}'.format(file_content))
+
+        if filetype.lower() in ['json']:
+            return json.loads(file_content)
+        elif filetype.lower() in ['yaml', 'yml']:
+            return yaml.safe_load(file_content)
+        else:
+            raise ConfigException(
+                'Unsupported filetype provided: {}'.format(filetype)
+            )
+
+    def load_config(self, config):
+        cp = ConfigProvider(config)
+
+        envs = self._load_as_dict_from_repo(
+            file_path=cp.envs_path,
+            repo_name=cp.envs_repo,
+            repo_branch=cp.envs_repo_branch,
+            filetype=cp.envs_filetype
+        )
+
+        secrets = self._load_as_dict_from_repo(
+            file_path=cp.secrets_path,
+            repo_name=cp.secrets_repo,
+            repo_branch=cp.secrets_repo_branch,
+            filetype=cp.secrets_filetype
+        )
+
+        return envs, secrets
+
+    def rotate(self, secret):
+        raise RotationException("Not implemented for GitLab")
+
+    @ exponential_backoff_retry(3)
+    def distribute(self, secret, destination):
+        return self._distribute(secret, destination)
+
+    @ classmethod
+    def validate_spec(cls, spec):
+        valid, msg = BaseProvider.validate_spec(spec)
+        if not valid:
+            return valid, msg
+
+        try:
+            SPEC_SCHEMA.validate(spec)
+            return True, "Validation succeeded."
+        except SchemaError as e:
+            print(e)
```

Comment:
Any reason for not using `LOGGER.warn` here?

Reply:
Oh, nice catch.